### PR TITLE
TCS-137 - Filtered Tailer Savedown on Timer

### PR DESCRIPTION
**Pull request by CONeillAQ**
Tailer functionality for only saving tables from memory that pass a configurable threshold. 
Check this condition on a configurable timer.
Improve savedown logging.
Reload TailReader tables from TDB.

### CODE REVIEW

File: code/processes/tailer.q [hunk 1]
```diff
@@ -50,9 +51,10 @@ getprocs:{[x;y]
...
-initdatastripe[];
-
+if[.ds.datastripe;
+  .lg.o[`dsinit;"datastripe on: initialising datastripe"];
+  initdatastripe[]];
 
 /- create HDB sym file and taildir symlink
-.ds.symlink[];
+if[.ds.datastripe;
+  .ds.symlink[]];
```
Comment (SeanW17) [inline code/processes/tailer.q:60]: Shouldn't be needed as tailer will only be on if .ds.datastripe is enabled in the STP.

File: code/processes/tailer.q [hunk 2]
```diff
@@ -50,9 +51,10 @@ getprocs:{[x;y]
 \d .
 
 /- initialise datastripe
-.lg.o[`dsinit;"initialising datastripe"];
-initdatastripe[];
-
+if[.ds.datastripe;
+  .lg.o[`dsinit;"datastripe on: initialising datastripe"];
+  initdatastripe[]];
```
Comment (SeanW17) [inline code/processes/tailer.q:56]: Shouldn't be needed as tailer will only be on if .ds.datastripe is enabled in the STP.

File: code/tailer/datastripe.q [hunk 3]
```diff
@@ -118,33 +112,40 @@ upserttopartition:{[dir;tablename;keycol;enumdata;nextp]
...
     };
 
-savealltablesoverperiod:{[dir;nextp;lasttime]
+savealltablesoverperiod:{[]
     /- function takes the tailer hdb directory handle and a timestamp
     /- saves each table up to given period to their respective partitions
-    savetablesoverperiod[dir;;nextp;lasttime]each .wdb.tablelist[];
+    dir:.ds.td; 
+    nextp:.z.p;
+    lasttime:.z.p-`second$.wdb.period;
```
Comment (SeanW17) [inline code/tailer/datastripe.q:134]: Do these need defined on separate lines or can they be done within the functions they are used in?
Do you need to substitute the names or can you just use .ds.td and .z.p in most lines?

File: code/tailer/datastripe.q [hunk 4]
```diff
@@ -118,33 +112,40 @@ upserttopartition:{[dir;tablename;keycol;enumdata;nextp]
...
+    totals:{count get x}each .wdb.tablelist[];
+    .wdb.timepartition:.z.p;
```
Comment (SeanW17) [inline code/tailer/datastripe.q:136]: Is this variable used?

File: code/tailer/datastripe.q [hunk 5]
```diff
@@ -118,33 +112,40 @@ upserttopartition:{[dir;tablename;keycol;enumdata;nextp]
...
+    if[max totals>.wdb.rowthresh;
+        .lg.o[`save;"Saving ",(", " sv string .wdb.tablelist[] where totals>.wdb.rowthresh)," table(s)"];
+        savetablesoverperiod[dir;;nextp;lasttime]each (.wdb.tablelist[] where totals>.wdb.rowthresh);
     /- trigger reload of access tables and intradayDBs in all tail reader processes
-    .tailer.dotailreload[`]};
+    .tailer.dotailreload[`]
+    ];  
+    if[min totals<.wdb.rowthresh;
+        .lg.o[`save;"No tables above threshold, no tables saved"]
```
Comment (SeanW17) [inline code/tailer/datastripe.q:144]: Where is rowthresh defined and should it be quoted in this comment?

File: code/tailer/datastripe.q [hunk 6]
```diff
@@ -118,33 +112,40 @@ upserttopartition:{[dir;tablename;keycol;enumdata;nextp]
...
-
     /- enumerate and then split by keycol
     symdir:` sv dir,.proc.procname;
     enumkeycol: .Q.en[symdir;?[tablename;enlist (<;`time;nextp);0b;()]];
     splitkeycol: {[enumkeycol;keycol;s] ?[enumkeycol;enlist (=;keycol;enlist s);0b;()]}[enumkeycol;keycol;] each partitionlist;
-
     /-upsert table to partition
-    upserttopartition[dir;tablename;keycol;;nextp] each splitkeycol where 0<count each splitkeycol; 
-
+    upserttopartition[dir;tablename;keycol;;nextp] each splitkeycol where 0<count each splitkeycol;
```
[the remark this answers is not in the document]
Comment (CONeillAQ) [reply code/tailer/datastripe.q:122]: Unsure as to why this is listed as a change, although it could have been a result of how a merge conflict was resolved. Regardless, this line works unchanged.

File: code/processes/tailer.q (hunk 1 above)
Comment (CONeillAQ) [reply code/processes/tailer.q:60]: No worries, will streamline this by the end of the day

File: code/processes/tailer.q (hunk 2 above)
Comment (CONeillAQ) [reply code/processes/tailer.q:56]: No worries, will streamline this by the end of the day

File: code/tailer/datastripe.q (hunk 3 above)
Comment (CONeillAQ) [reply code/tailer/datastripe.q:134]: The names were substituted to improve readability but it should work without the replacement.

File: code/tailer/datastripe.q (hunk 4 above)
Comment (aidanloughranAQ) [reply code/tailer/datastripe.q:136]: That seems to be a variable we don't use anymore so we can take this out

File: code/tailer/datastripe.q (hunk 5 above)
Comment (CONeillAQ) [reply code/tailer/datastripe.q:144]: rowthresh is defined in the tailer settings file, will add a comment now

File: code/processes/tailer.q [hunk 7]
```diff
@@ -51,10 +51,8 @@ getprocs:{[x;y]
...
-  .lg.o[`dsinit;"datastripe on: initialising datastripe"];
-  initdatastripe[]];
+.lg.o[`dsinit;"initialising datastripe"];
+initdatastripe[];
 
 /- create HDB sym file and taildir symlink
-if[.ds.datastripe;
-  .ds.symlink[]];
\ No newline at end of file
+.ds.symlink[]];
```
Comment (SeanW17) [inline code/processes/tailer.q]: Think you still have an extra right square bracket.

Comment (SeanW17) [reply code/processes/tailer.q]: Nevermind seen your next commit.

**Review by SeanW17 — APPROVED** [review]

File: code/tailer/datastripe.q [hunk 8]
```diff
@@ -118,33 +112,40 @@ upserttopartition:{[dir;tablename;keycol;enumdata;nextp]
...
+    ];
+    /- if totals are below rowthresh log message indicates no tables saved
+    if[min totals<.wdb.rowthresh;
+        .lg.o[`save;"No tables above threshold, no tables saved"]
+    ];
+    };
 
-.timer.repeat[00:00+.z.d;0W;0D00:10:00;(`.ds.savealltablesoverperiod;.ds.td;.z.p);"Saving tables"]
+/- Timer to repeat savealltablesoverperiod with period defined in tailer.q settings
+.timer.repeat[00:00+.z.d;0W;.wdb.period;(`.ds.savealltablesoverperiod;`);"Saving tables"]
```
Comment (jonathonmcmurray) [inline code/tailer/datastripe.q:148]: should we be using `.wdb.period` here? I think the "typical" period that will be used is 1 hour, so we'll only check once per hour if we've met the threshold?

how often does the regular wdb check this?

File: config/settings/tailer.q [hunk 9]
```diff
@@ -50,7 +50,9 @@ eodwaittime:0D00:00:10.000
 tpcheckcycles:0W                                                                            // number of attempts to connect to tp before process is killed
 
 \d .ds
+datastripe:0b
 period:0D01
+lasttimestamp:.z.p-4*period
```
Comment (jonathonmcmurray) [inline config/settings/tailer.q:56]: are these changes relevant to this PR? I don't understand why they're here

File: tests/tailer/savedown_withcount/savedown.csv [hunk 10]
```diff
@@ -0,0 +1,20 @@
+action,ms,bytes,lang,code,repeate,minver,comment
+before,0,0,q,tr:.servers.gethandlebytype[`tr_seg1;`any],1,,"Connect to tailreader handle to give access the tailreader1 for testing tailer1"
+before,0,0,q,tl:.servers.gethandlebytype[`tailer_seg1;`any],1,,"Connect to tailer1 handle,so tailer can be tested through tl handle"
+run,0,0,q,(tr"count tables[]")=2,1,,"Check the number of tables in the tailreader is currently 2"
+run,0,0,q,tl".ds.savealltablesoverperiod[]",1,,"Running savedown function sould save tables but, there should be no data to savedown so the table count should stay the same"
+run,0,0,q,(tr"count tables[]")=2,1,,"Check the number of tables in the tailreader has remained the same indicating the savedown didn't add any tables as there was none to add"
+run,0,0,q,system "sleep 10",1,,"Wait 10 seconds to populate tables with rows of data for trade and quote"
```
Comment (jonathonmcmurray) [inline tests/tailer/savedown_withcount/savedown.csv:7]: torq has .os.sleep

```suggestion
run,0,0,q,.os.sleep 10,1,,"Wait 10 seconds to populate tables with rows of data for trade and quote"
```

File: code/tailer/datastripe.q [hunk 11]
```diff
@@ -118,33 +112,40 @@ upserttopartition:{[dir;tablename;keycol;enumdata;nextp]
...
 
-savealltablesoverperiod:{[dir;nextp;lasttime]
+savealltablesoverperiod:{[]
     /- function takes the tailer hdb directory handle and a timestamp
     /- saves each table up to given period to their respective partitions
-    savetablesoverperiod[dir;;nextp;lasttime]each .wdb.tablelist[];
+    /- totals calculates the row count of tables in .wdb.tablelist[] 
+    totals:{count get x}each .wdb.tablelist[];
+    /- If a totals value is greater than rowthresh which is the rowcount for saving down tables
+    if[max totals>.wdb.rowthresh;
```
Comment (jonathonmcmurray) [inline code/tailer/datastripe.q:135]: IMO this should be either

```suggestion
    if[max[totals]>.wdb.rowthresh;
```

or

```suggestion
    if[any totals>.wdb.rowthresh;
```

Technically these will all serve the same purpose, but I think it is semantically clearer to use `any` here

File: code/tailer/datastripe.q [hunk 12]
```diff
@@ -118,33 +112,40 @@ upserttopartition:{[dir;tablename;keycol;enumdata;nextp]
...
     .gc.run[];
     };
 
-savealltablesoverperiod:{[dir;nextp;lasttime]
+savealltablesoverperiod:{[]
     /- function takes the tailer hdb directory handle and a timestamp
     /- saves each table up to given period to their respective partitions
-    savetablesoverperiod[dir;;nextp;lasttime]each .wdb.tablelist[];
+    /- totals calculates the row count of tables in .wdb.tablelist[] 
+    totals:{count get x}each .wdb.tablelist[];
```
Comment (jonathonmcmurray) [inline code/tailer/datastripe.q:133]: `get` is for loading tables from disk, I think `value` is more correct here (functionally identical but conventionally `value` is used to evaluate a symbol/string, so it's clearer what's happening in this line if we use `value`)

```suggestion
    totals:{count value x}each .wdb.tablelist[];
```

File: code/tailer/datastripe.q [hunk 13]
```diff
@@ -118,33 +112,40 @@ upserttopartition:{[dir;tablename;keycol;enumdata;nextp]
...
+    /- If a totals value is greater than rowthresh which is the rowcount for saving down tables
+    if[max totals>.wdb.rowthresh;
+        .lg.o[`save;"Saving ",(", " sv string .wdb.tablelist[] where totals>.wdb.rowthresh)," table(s)"];
+        savetablesoverperiod[.ds.td;;.z.p;(.z.p-`second$.wdb.period)]each (.wdb.tablelist[] where totals>.wdb.rowthresh);
     /- trigger reload of access tables and intradayDBs in all tail reader processes
-    .tailer.dotailreload[`]};
+    .tailer.dotailreload[`]
+    ];
+    /- if totals are below rowthresh log message indicates no tables saved
+    if[min totals<.wdb.rowthresh;
```
Comment (jonathonmcmurray) [inline code/tailer/datastripe.q:142]: as above but

```suggestion
    if[not any totals<.wdb.rowthresh;
```

or

```suggestion
    if[all totals>.wdb.rowthresh;
```

I would also move this conditional to the start of the function & return early from it (i.e. `:();`), then have the code that does the save *outside* any conditional (because obviously if the first conditional is false, the other is true, so don't necessarily need to specify it)

**Review by jonathonmcmurray — COMMENTED** [review]
Please make sure to have a descriptive PR title - it's not clear to me what the current title means

There seem to be a few unrelated/non-functional changes in the PR - always best to keep the PR as clean as possible with only the changes that are actually relevant, to make it as easy as possible to review

On a larger point - the wdb already has the functionality to check row thresholds on a timer - is there any way we can use the existing functionality or do we _need_ to reimplement it here?

File: code/tailer/datastripe.q (hunk 12 above)
Comment (CONeillAQ) [reply code/tailer/datastripe.q]: This has been updated to use `value`

File: code/tailer/datastripe.q (hunk 13 above)
Comment (CONeillAQ) [reply code/tailer/datastripe.q]: This has been restructured to only use one conditional, the semantics of which has been improved as per this suggestion to use `all`

File: code/tailer/datastripe.q (hunk 11 above)
Comment (CONeillAQ) [reply code/tailer/datastripe.q]: This conditional has been removed, and the function has been restructured to only use one conditional

File: tests/tailer/savedown_withcount/savedown.csv (hunk 10 above)
Comment (CONeillAQ) [reply tests/tailer/savedown_withcount/savedown.csv:7]: This has been fixed

File: config/settings/tailer.q (hunk 9 above)
Comment (CONeillAQ) [reply config/settings/tailer.q]: These lines have now been removed

**Review by eimear-dalzell — APPROVED** [review]

File: code/tailer/datastripe.q [hunk 14]
```diff
@@ -89,22 +89,16 @@ upserttopartition:{[dir;tablename;keycol;enumdata;nextp]
     /- function takes a (dir)ectory handle, tablename as a symbol
     /- column to key table on, an enumerated table and a timestamp.
     /- partitions the data on the keycol and upserts it to the given dir
-
```
[the remark this answers is not in the document]
Comment (jonathonmcmurray) [reply code/tailer/datastripe.q:109]: this was marked as resolved, but it's still true that there are non-functional whitespace changes

File: code/tailer/datastripe.q [hunk 15]
```diff
@@ -151,17 +151,26 @@ savetables:{[dir;tablename]
...
+    totals:{count value x}each .wdb.tablelist[];
+    /- log and return from function early if no table has crossed threshold
+    if[all totals<.wdb.numrows;
+        .lg.o[`save;"No tables above threshold, no tables saved"];
+        :();
+        ];
+
+    /- log and savedown any tables above threshold
+    .lg.o[`save;"Saving ",(", " sv string .wdb.tablelist[] where totals>.wdb.numrows)," table(s)"];
+    savetables[dir;]each (.wdb.tablelist[] where totals>.wdb.numrows);
```
Comment (jonathonmcmurray) [inline code/tailer/datastripe.q:167]:
```suggestion
    .lg.o[`save;"Saving ",(", " sv string tabstosave:.wdb.tablelist[] where totals>.wdb.numrows)," table(s)"];
    savetables[dir;]each tabstosave;
```

Don't need to derive same list of tables twice

File: code/tailer/datastripe.q [hunk 16]
```diff
@@ -151,17 +151,26 @@ savetables:{[dir;tablename]
...
-    /- delete data that has been saved
-    @[`.;;0#] each .wdb.tablelist[];
-	
+    /- totals calculates the row count of tables in .wdb.tablelist[]
+    totals:{count value x}each .wdb.tablelist[];
+    /- log and return from function early if no table has crossed threshold
+    if[all totals<.wdb.numrows;
+        .lg.o[`save;"No tables above threshold, no tables saved"];
+        :();
+        ];
```
Comment (jonathonmcmurray) [inline code/tailer/datastripe.q:163]:
```suggestion
    ];
```

IMO the closing `]` of a multi-line `if` should be aligned with the `i` in opening line, just makes it clearer when scanning the file/function quickly

File: code/tailer/datastripe.q [hunk 17]
```diff
@@ -151,17 +151,26 @@ savetables:{[dir;tablename]
     .gc.run[];
     };
 
-savealltables:{[dir]
+savealltablesoverperiod:{[dir;lasttime]
```
Comment (jonathonmcmurray) [inline code/tailer/datastripe.q:154]: I can't see where `lasttime` is used, am I just being blind?

File: tests/tailer/savedown_withcount/settings.q [hunk 18]
```diff
@@ -0,0 +1,9 @@
+// IPC connection parameters
+.servers.CONNECTIONS:`segmentedtickerplant`sort`gateway`rdb`hdb`tailer_seg1`tr_seg1
+.servers.USERPASS:`admin:admin;
```
Comment (jonathonmcmurray) [inline tests/tailer/savedown_withcount/settings.q:3]: the test proc has the `tailer` proctype, so surely it should be automatically using the tailer username & pass? i.e. this line shouldn't be needed I don't think

File: tests/tailer/savedown_withcount/tests.csv [hunk 19]
```diff
@@ -0,0 +1,16 @@
+action,ms,bytes,lang,code,repeate,minver,comment
+before,0,0,q,tr:.servers.gethandlebytype[`tr_seg1;`any],1,,"Connect to tailreader handle to give access the tailreader1 for testing tailer1"
+before,0,0,q,tl:.servers.gethandlebytype[`tailer_seg1;`any],1,,"Connect to tailer1 handle,so tailer can be tested through tl handle"
+before,0,0,q,tl".wdb.numrows:100",1,,"Set row count as the standard row count is 10000 would be to large for a test"
+run,0,0,q,(tr"count tables[]")=2,1,,"Check the number of tables in the tailreader is currently 2"
+run,0,0,q,tl".ds.savealltablesoverperiod[]",1,,"Running savedown function sould save tables but, there should be no data to savedown so the table count should stay the same"
```
Comment (jonathonmcmurray) [inline tests/tailer/savedown_withcount/tests.csv:6]: if I'm not mistaken, this function takes two parameters? how does this test work if that's the case?

File: tests/tailer/savedown_withcount/tests.csv [hunk 20]
```diff
@@ -0,0 +1,16 @@
...
+run,0,0,q,(tr"count tables[]")=2,1,,"Check the number of tables in the tailreader has remained the same indicating the savedown didn't add any tables as there was none to add"
+run,0,0,q,.os.sleep 10,1,,"Wait 10 seconds to populate tables with rows of data for trade and quote"
+run,0,0,q,tl"![;();0b;`symbol$()]`trade";tl".ds.savealltablesoverperiod[]",1,,"Running savedown function on tables,should save quote"
```
Comment (jonathonmcmurray) [inline tests/tailer/savedown_withcount/tests.csv:9]: Why the functional form here? Surely we can just do this?

```suggestion
run,0,0,q,tl"delete from `trade";tl".ds.savealltablesoverperiod[]",1,,"Running savedown function on tables,should save quote"
```

it's also not explained in the comment why we're deleting from trade table

File: tests/tailer/savedown_withcount/tests.csv [hunk 21]
```diff
@@ -0,0 +1,16 @@
+action,ms,bytes,lang,code,repeate,minver,comment
+before,0,0,q,tr:.servers.gethandlebytype[`tr_seg1;`any],1,,"Connect to tailreader handle to give access the tailreader1 for testing tailer1"
+before,0,0,q,tl:.servers.gethandlebytype[`tailer_seg1;`any],1,,"Connect to tailer1 handle,so tailer can be tested through tl handle"
+before,0,0,q,tl".wdb.numrows:100",1,,"Set row count as the standard row count is 10000 would be to large for a test"
+run,0,0,q,(tr"count tables[]")=2,1,,"Check the number of tables in the tailreader is currently 2"
```
Comment (jonathonmcmurray) [inline tests/tailer/savedown_withcount/tests.csv:5]: this seems very config dependent etc., I'm not sure we should be failing a test if e.g. there are more than 2 tables?

**Review by jonathonmcmurray — COMMENTED** [review]
few questions here, particularly around the tests

File: code/tailer/datastripe.q (hunk 17 above)
Comment (CONeillAQ) [reply code/tailer/datastripe.q:154]: I believe this is just a vestige of how the function used to operate, this can be safely removed

File: code/tailer/datastripe.q [hunk 22]
```diff
@@ -151,17 +151,26 @@ savetables:{[dir;tablename]
     .gc.run[];
     };
 
-savealltables:{[dir]
+savealltablesoverperiod:{[dir]
```
[the remark this answers is not in the document]
Comment (ThomasGillespie) [reply code/tailer/datastripe.q:154]: other than that, looks good

[the remark this answers is not in the document]
Comment (CONeillAQ) [reply code/tailer/datastripe.q]: I have now updated this branch to be more compliant with the changes made in #514

File: code/tailer/datastripe.q [hunk 23]
```diff
@@ -154,14 +154,26 @@ savetables:{[dir;tablename]
 savealltables:{[dir]
     /- function takes the tailer hdb directory handle and a timestamp
     /- saves each table up to given period to their respective partitions
-    savetables[dir;]each .wdb.tablelist[];
-	
+    /- totals calculates the row count of tables in .wdb.tablelist[]
+    totals:{count value x}each .wdb.tablelist[];
+    /- log and return from function early if no table has crossed threshold
+    if[all totals<.wdb.numrows;
```
Comment (dmitchard) [inline code/tailer/datastripe.q:160]: As the savealltables function is used by both eod and eop, this condition will cause issues at eop and eod where the save actually might not occur because of it. I think this needs to be seperated into its own function outside of savealltables so that the functionality is just for saving and nothing else.

**Review by dmitchard — CHANGES_REQUESTED** [review]
As the savealltables function is used by both eod and eop, this condition will cause issues at eop and eod where the save actually might not occur because of it. I think this needs to be seperated into its own function outside of savealltables so that the functionality is just for saving and nothing else.

**Review by ThomasGillespie — APPROVED** [review]

**Review by dmitchard — APPROVED** [review]

**Review by jonathonmcmurray — APPROVED** [review]
